pkg/manager/volumes/vol_compare_utils.go

```diff
@@ -0,0 +1,371 @@
+	return &volCompareUtils{
+		deps: deps,
+		sf:   MustNewSelectorFactory(),
+	}
+}
+
+type componentVolumeContext struct {
+	context.Context
+	tc     *v1alpha1.TidbCluster
+	status v1alpha1.ComponentStatus
+
+	shouldEvict bool
+
+	pods []*corev1.Pod
+	sts  *appsv1.StatefulSet
+
+	desiredVolumes []DesiredVolume
+}
+
+func (c *componentVolumeContext) ComponentID() string {
+	return fmt.Sprintf("%s/%s:%s", c.tc.GetNamespace(), c.tc.GetName(), c.status.MemberType())
+}
+
+func (u *volCompareUtils) BuildContextForTC(tc *v1alpha1.TidbCluster, status v1alpha1.ComponentStatus) (*componentVolumeContext, error) {
+	comp := status.MemberType()
+
+	ctx := &componentVolumeContext{
+		Context: context.TODO(),
+		tc:      tc,
+		status:  status,
+	}
+
+	vs, err := u.GetDesiredVolumes(tc, comp)
+	if err != nil {
+		return nil, err
+	}
+	ctx.desiredVolumes = vs
+
+	sts, err := u.getStsOfComponent(tc, comp)
+	if err != nil {
+		return nil, err
+	}
+
+	pods, err := u.getPodsOfComponent(tc, comp)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.pods = pods
+	ctx.sts = sts
+	ctx.shouldEvict = comp == v1alpha1.TiKVMemberType
+
+	return ctx, nil
+}
+
+func (u *volCompareUtils) getPodsOfComponent(tc *v1alpha1.TidbCluster, mt v1alpha1.MemberType) ([]*corev1.Pod, error) {
+	selector, err := u.sf.NewSelector(tc.GetInstanceName(), mt)
+	if err != nil {
+		return nil, err
+	}
+
+	ns := tc.GetNamespace()
+
+	pods, err := u.deps.PodLister.Pods(ns).List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list Pods: %w", err)
+	}
+
+	sort.Slice(pods, func(i, k int) bool {
+		a, b := pods[i].Name, pods[k].Name
+		if len(a) != len(b) {
+			return len(a) < len(b)
+		}
+		return a < b
+	})
+
+	return pods, nil
+}
+
+func (u *volCompareUtils) getStsOfComponent(cluster v1alpha1.Cluster, mt v1alpha1.MemberType) (*appsv1.StatefulSet, error) {
+	ns := cluster.GetNamespace()
+	stsName := controller.MemberName(cluster.GetName(), mt)
+
+	sts, err := u.deps.StatefulSetLister.StatefulSets(ns).Get(stsName)
+	if err != nil {
+		return nil, fmt.Errorf("get sts %s/%s failed: %w", ns, stsName, err)
+	}
+
+	return sts, nil
+}
+
+func (u *volCompareUtils) IsStatefulSetSynced(ctx *componentVolumeContext, sts *appsv1.StatefulSet) (bool, error) {
+	// Note: The only errors that IsStatefulSetSynced returns is when number of disks change. This cannot be handled by
+	// pvc_resizer so is an error for it, however pvc_replacer can handle it and will ignore all errors returned here
+	// and only rely on the boolean return (which must be set to false if number of disks change on top of error)
+	for _, volTemplate := range sts.Spec.VolumeClaimTemplates {
+		volName := v1alpha1.StorageVolumeName(volTemplate.Name)
+		size := getStorageSize(volTemplate.Spec.Resources.Requests)
+		desired := getDesiredVolumeByName(ctx.desiredVolumes, volName)
+		if desired == nil {
+			return false, fmt.Errorf("volume %s in sts for cluster %s does not exist in desired volumes", volName, ctx.ComponentID())
```

Comment:
May this returned error block the reconciliation if the additional volume is used?

Reply:
Please read the Note in comments above (~line 153).
This function is used by both pvc replacer and pvc modifier. In replacer error is ignored because differences will be fixed by it. For pvc modifier it is intentional error that we should block reconciliation.

If user does not have replacer enabled and has requested additional volume, pvc modifier cannot make the change and I feel better to explicitly block reconciliation with error than silently ignoring additional volume and rolling out half a change for user.

Reply:
But if the user does not want to use the volume replace feature, the previous volume modify feature will be blocked here. If other words, the previous behavior is broken.

Reply:
Yes this is actually a change in existing behavior, but I believe the existing behavior was also problematic.

I'd called this out earlier in review here: https://github.com/pingcap/tidb-operator/pull/5150/commits/4f3ef146cb15d16861e05ecc2de255414f785f34#r1275584130
And wasn't sure which way pingcap folks preferred.

Today pvc modifer would silently ignore extra volumes, which can be problematic. Fundamentally tidb operator does not have proper support for changing number of volumes before this PR.

Can you provide some use case when replacer feature is disabled that users will want to add a new volume?
New volume cannot be used without a config change. Today operator will push config change without the volume (silently ignoring volume) and will cause pods to crash because missing volume/path. Due to this user cannot make new volume anyways today, so I thought safer to block the reconciliation instead of silently ignoring it and causing pod crashes.

If you feel better to keep old behavior or have some use case where you think it's needed, I can revert to old behavior, let me know.

Reply:
#5011 may be one use case.

Reply:
It doesn't seem like the user was trying to change additionalMounts in that issue. Existing additionalMounts won't be a problem in this PR, changing additionalMounts is what is problematic.

I think we should decide what is the expected behavior if user changes additionalMount for an existing cluster without enabling replace feature. Today's behavior is to ignore the mount but rollout everything else , which I believe is rarely user's intent and can cause issues. Proposed behavior is to explicitly block upgrade as it is explicitly unsupported instead of doing a partial update

Either case, should not affect the replace feature. So if you prefer old behavior I can split this function and revert for existing modifier's behavior. Let me know.

Reply:
If the user changes the additionalMount without enabling replace feature, then TiDB Operator blocks it should be OK.

But if the user changes the main PVC (e.g size) but with additionalMount, then TiDB Operator should not block when checking this additionalMount.

Reply:
That should still be fine here?
This line only blocks if desired = nil, but a volume was defined in current spec. (I.e storageVolumes was defined).
Note that additionalVolume/mounts are not used anywhere in this logic, only storageVolumes field.

If additional volumes/mounts or storage volumes are used , this code won't block as long as they are not being changed. If only size is being changed line 163 below will catch it return a difference without error thereby not blocking it.

Reply:
`additionalVolumes` are append into `sts.spec.podTemplate.spec.volumes` but not `sts.spec.volumeClaimTemplates`.

The `storageVolume` deletion is not well handled in previous logic. There are two problems:
1. STS will always be recreated if it's not synced. Some PVCs maybe left if `storageVolumes` are deleted.
2. If only `storageVolume` are deleted, STS is synced.

There are three choice to fix this behavior:
1. Throw a clear err to notify users that changes of `storageVolumes` is unsupported if this feature is not enabled.
2. Fix both problem 1 and 2, clean up left PVCs of deleted `storageVolumes`.
3. Fix the problem 2. And just announce that PVCs of deleted `storageVolumes` will be left because of safety.
 
For me, option 1 > 3 > 2. So I think it's enough to add a more clear comment and return a more clear error.
@csuzhangxc What do you think about?

Reply:
@liubog2008 Thanks for your detailed explanation, I agree with your suggestion ( 1 > 3 > 2).
@anish-db How about making the error messages more explicit as it's unsupported?

Reply:
Sure, I've expanded the error warning to be more prominent. I've added this under pvc_modifier.go where this function gets used and the error is handled. New Error reads like:

`change in number of volumes is not supported without VolumeReplacing feature. Reconciliation is blocked due to :`

Followed by either of the two error conditions returned by this function.